execute DDL statement in transaction reports 'invalid transaction'

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists t;
create table t (c_int int, c_str varchar(40));
insert into t values (1, 'quizzical hofstadter');
begin;
select c_int from t where c_str is not null for update;
alter table t add index idx_4 (c_str);
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

```console
Query OK, 0 rows affected
```

### 3. What did you see instead (Required)

```console
ERROR 8024 (HY000): invalid transaction
```

log output:

```
[2020/12/14 19:19:13.332 +08:00] [ERROR] [txn.go:215] ["the code should never run here"] [TxnState="Txn{state=valid, txnStartTS=421513069415104512}"] ["staging handler"=1] ["something must be wrong"="github.com/pingcap/tidb/session.(*TxnState).Commit\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/txn.go:218\ngithub.com/pingcap/tidb/session.(*session).doCommit\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:478\ngithub.com/pingcap/tidb/session.(*session).doCommitWithRetry\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:499\ngithub.com/pingcap/tidb/session.(*session).CommitTxn\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:559\ngithub.com/pingcap/tidb/session.(*session).NewTxn\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:1627\ngithub.com/pingcap/tidb/executor.(*DDLExec).Next\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/executor/ddl.go:77\ngithub.com/pingcap/tidb/executor.Next\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/executor/executor.go:278\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/executor/adapter.go:524\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/executor/adapter.go:405\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/executor/adapter.go:355\ngithub.com/pingcap/tidb/session.runStmt\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:1310\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/session/session.go:1254\ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/driver_tidb.go:212\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/conn.go:1561\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/conn.go:1453\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/conn.go:1027\ngithub.com/pingcap/tidb/server.(*clientConn).Run\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/conn.go:792\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/tangenta/gopath/src/github.com/pingcap/tidb/server/server.go:461"]
```

### 4. What is your TiDB version? (Required)
master, `49b926ede766b316cc42bdf40d9ada37deb67968`

<!-- Paste the output of SELECT tidb_version() -->

--- a/executor/ddl_test.go
+++ b/executor/ddl_test.go
@@ -75,6 +75,17 @@ func (s *testSuite6) TestInTxnExecDDLFail(c *C) {
 	result.Check(testkit.Rows("1"))
 }
 
+func (s *testSuite6) TestInTxnExecDDLInvalid(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (c_int int, c_str varchar(40));")
+	tk.MustExec("insert into t values (1, 'quizzical hofstadter');")
+	tk.MustExec("begin;")
+	_ = tk.MustQuery("select c_int from t where c_str is not null for update;")
+	tk.MustExec("alter table t add index idx_4 (c_str);")
+}
+
 func (s *testSuite6) TestCreateTable(c *C) {
 	tk := testkit.NewTestKit(c, s.store)
 	tk.MustExec("use test")

--- a/session/session.go
+++ b/session/session.go
@@ -1309,21 +1309,6 @@ func runStmt(ctx context.Context, se *session, s sqlexec.Statement) (rs sqlexec.
 	}
 	rs, err = s.Exec(ctx)
 	sessVars.TxnCtx.StatementCount++
-	if !s.IsReadOnly(sessVars) {
-		// All the history should be added here.
-		if err == nil && sessVars.TxnCtx.CouldRetry {
-			GetHistory(se).Add(s, sessVars.StmtCtx)
-		}
-
-		// Handle the stmt commit/rollback.
-		if se.txn.Valid() {
-			if err != nil {
-				se.StmtRollback()
-			} else {
-				se.StmtCommit()
-			}
-		}
-	}
 	if rs != nil {
 		return &execStmtResult{
 			RecordSet: rs,

--- a/session/tidb.go
+++ b/session/tidb.go
@@ -197,6 +197,22 @@ func recordAbortTxnDuration(sessVars *variable.SessionVars) {
 }
 
 func finishStmt(ctx context.Context, se *session, meetsErr error, sql sqlexec.Statement) error {
+	sessVars := se.sessionVars
+	if !sql.IsReadOnly(sessVars) {
+		// All the history should be added here.
+		if meetsErr == nil && sessVars.TxnCtx.CouldRetry {
+			GetHistory(se).Add(sql, sessVars.StmtCtx)
+		}
+
+		// Handle the stmt commit/rollback.
+		if se.txn.Valid() {
+			if meetsErr != nil {
+				se.StmtRollback()
+			} else {
+				se.StmtCommit()
+			}
+		}
+	}
 	err := autoCommitAfterStmt(ctx, se, meetsErr, sql)
 	if se.txn.pending() {
 		// After run statement finish, txn state is still pending means the

